add lecture 02 slides

andrei-shulaev opened the pull request.

ashulaev committed:
add slides for lecture 02

Cjkjvfnby commented:
Diff is not shown in GitHub, too many lines. See `To see what is in scope` section.

Please capitalize new lines.
- ![image](https://user-images.githubusercontent.com/171597/97631700-67688980-1a42-11eb-8e6c-f474d2c1cf82.png)
- ![image](https://user-images.githubusercontent.com/171597/97633865-cf6c9f00-1a45-11eb-8b15-90346dd36534.png)
- ![image](https://user-images.githubusercontent.com/171597/97635137-eb714000-1a47-11eb-9f96-4659c4f78080.png)


> Mutate string

Confusing title for creating a new instance of the string. 

> String formatting

Should we talk about formating arguments https://docs.python.org/3/library/string.html#formatspec?

> Nice alternative to map and filter

I'd reccoment skiping the part with map and filter.

Looks like github shows me the same image for all pictures.
![image](https://user-images.githubusercontent.com/171597/97634891-89183f80-1a47-11eb-84ed-3a5603480505.png)

> This behavior could be canged global and nonlocal operators

Missed preposition

cdeler reviewed (commented):
As the diff is too large, I left the comments there:

1. I'd mention `NoneType` in this diagram "A mutable object can be changed after it's created, and an immutable object can't"
1. From my point of view it's better not to use `**kwargs` in the presentation (or to show what's that)
1. I'd format all the code in cells using `black` (using [blackcellmagic](https://github.com/csurfer/blackcellmagic) or [nbQA](https://github.com/nbQA-dev/nbQA))

andrei-shulaev commented:
> Diff is not shown in GitHub, too many lines. See `To see what is in scope` section.

removed 
> Please capitalize new lines.

done
> Confusing title for creating a new instance of the string.

changed
> Should we talk about formating arguments https://docs.python.org/3/library/string.html#formatspec?
 
> > Nice alternative to map and filter
> 
> I'd reccoment skiping the part with map and filter.

Not agree. I'm sure students will have strong intention to overuse functional approach after FP lecture. They should know that using of list comprehensions is preferable. 

> Looks like github shows me the same image for all pictures.
> ![image](https://user-images.githubusercontent.com/171597/97634891-89183f80-1a47-11eb-84ed-3a5603480505.png)

Some images are simply copy-pasted into the cell. So they have the same attachments name. It should work fine locally.

> Missed preposition

Changed

andrei-shulaev commented:
>     1. I'd mention `NoneType` in this diagram "A mutable object can be changed after it's created, and an immutable object can't"
 
Will do it.

>     2. From my point of view it's better not to use `**kwargs` in the presentation (or to show what's that)

I do it when explaining packing/unpacking using double asterisk

>     3. I'd format all the code in cells using `black` (using [blackcellmagic](https://github.com/csurfer/blackcellmagic) or [nbQA](https://github.com/nbQA-dev/nbQA))

I would prefer to stay as is. Formatter fails on some cells and I have to make them shorter since I plan to show it in presentation mode.

Cjkjvfnby commented:
>  I'm sure students will have strong intention to overuse functional approach after FP lecture. They should know that using of list comprehensions is preferable.

Then let's move it to the FP lecture.

Cjkjvfnby commented:
> Some images are simply copy-pasted into the cell. So they have the same attachments name. It should work fine locally.

That the reason for the huge diff.
```
 "attachments": {
    "image.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtAAAAIGCAIA..............
```

andrei-shulaev commented:
> Then let's move it to the FP lecture.

Removed. We have to make sure it would be mentioned during next lesson.

andrei-shulaev commented:
> That the reason for the huge diff.

Now it is viewable.

cdeler commented on lecture_02_data_structures_functions/hw/hw2.py:
```diff
@@ -0,0 +1,22 @@
+"""
+Given an array of size n, find the max and min element.
+The max element is the element that appears more than ⌊ n/2 ⌋ times.
```
May be rewrite it in python terms: 

```suggestion
The max element is the element that appears more than  n // 2 times.
```

cdeler reviewed (dismissed):
Approved in advance

Cjkjvfnby commented on lecture_02_data_structures_functions/hw/hw1.py:
```diff
@@ -0,0 +1,29 @@
+from typing import List
+
+
+def get_longest_diverse_words(file_path: str) -> List[str, 10]:
```
I haven't found any mention of such a size declaration in typing.List. 

```suggestion
def get_longest_diverse_words(file_path: str) -> List[str]:
```

Cjkjvfnby commented on lecture_02_data_structures_functions/hw/hw2.py:
```diff
@@ -0,0 +1,22 @@
+
+Input: [3,2,3]
+Output: 3, 2
+Example 2:
```
```suggestion
Example 2:

````

Cjkjvfnby commented on lecture_02_data_structures_functions/hw/hw2.py:
```diff
@@ -0,0 +1,22 @@
+"""
+Given an array of size n, find the max and min element.
```
```suggestion
Given an array of size n, find the most common and the least common elements.
```

Cjkjvfnby commented on lecture_02_data_structures_functions/hw/hw2.py:
```diff
@@ -0,0 +1,22 @@
+"""
+Given an array of size n, find the max and min element.
+The max element is the element that appears more than n // times.
```
What is `times`?

cdeler replied:
```suggestion
The most common is the element that appears more than n // 2 times.
```

Cjkjvfnby commented on lecture_02_data_structures_functions/hw/hw2.py:
```diff
@@ -0,0 +1,22 @@
+from typing import List, Tuple
+
+
+def major_and_minor_elem(inp: List) -> Tuple[int, 2]:
```
```suggestion
def major_and_minor_elem(inp: List[int]) -> Tuple[int, int]:
```
https://docs.python.org/3/library/typing.html#typing.Tuple

cdeler commented on lecture_02_data_structures_functions/hw/hw2.py:
```diff
@@ -0,0 +1,22 @@
+"""
+Given an array of size n, find the max and min element.
+The max element is the element that appears more than n // times.
+The min element is the element that appears fewer than other.
```
```suggestion
The least common is the element that appears fewer than other.
```

cdeler commented on lecture_02_data_structures_functions/hw/hw2.py:
```diff
@@ -0,0 +1,22 @@
+"""
+Given an array of size n, find the max and min element.
+The max element is the element that appears more than n // times.
+The min element is the element that appears fewer than other.
+
+You may assume that the array is non-empty and the max element always exist in the array.
```
```suggestion
The least common is the element that appears fewer than other.

You may assume that the array is non-empty and the most common element always exists in the array.
```

cdeler commented on lecture_02_data_structures_functions/hw/hw4.py:
```diff
@@ -0,0 +1,25 @@
+"""
+Write a function that accept another function as an argument. Then it
```
```suggestion
Write a function that accepts another function as an argument. Then it
```

cdeler commented on lecture_02_data_structures_functions/hw/hw2.py:
```diff
@@ -0,0 +1,24 @@
+"""
+Given an array of size n, find the most common and the least common elements.
+The most common element is the element that appears more than n // 2 times.
+The least element is the element that appears fewer than other.
```
```suggestion
The least common element is the element that appears fewer than other.
```

cdeler committed:
Update lecture_02_data_structures_functions/hw/hw2.py

ashulaev committed:
Minor updates
